src/gasEstimator/gasLimitEstimator.ts

```diff
@@ -0,0 +1,24 @@
+import { Transaction } from "../core";
+
+interface ITransactionCostResponse {
+    gasLimit: number;
+}
+
+interface INetworkProvider {
+    estimateTransactionCost(tx: Transaction): Promise<ITransactionCostResponse>;
+}
+
+export class GasLimitEstimator {
+    private networkProvider: INetworkProvider;
+    private gasMultiplier: number;
+
+    constructor(networkProvider: INetworkProvider, gasMultiplier?: number) {
+        this.networkProvider = networkProvider;
+        this.gasMultiplier = gasMultiplier || 1;
+    }
+
+    async estimateGasLimit(transaction: Transaction): Promise<bigint> {
+        const gasLimit = (await this.networkProvider.estimateTransactionCost(transaction)).gasLimit;
+        return BigInt(gasLimit * this.gasMultiplier);
```

Comment:
We have to properly handle rational numbers e.g. do some rounding. Otherwise, `BigInt` will fail:

```
> BigInt(123.456)
Uncaught:
RangeError: The number 123.456 cannot be converted to a BigInt because it is not an integer
```

Reply:
done. now using `Math.floor()`.